contracts/RewardController.sol

```diff
@@ -0,0 +1,272 @@
+        if (block.number < startBlock) {
+            epochRewardPerBlock = _epochRewardPerBlock;
+        }  else {
+            uint256 nextEpoch = (block.number - startBlock) / epochLength + 1;
+            // if rewards are ongoing, update the future rewards but keep past and current
+            for (; nextEpoch < NUMBER_OF_EPOCHS; nextEpoch++) {
+                epochRewardPerBlock[nextEpoch] = _epochRewardPerBlock[nextEpoch];
+            }
+        }
+        emit SetEpochRewardPerBlock(epochRewardPerBlock);
+    }
+
+    function _setAllocPoint(address _vault, uint256 _allocPoint) internal {
+        if (vaultRewardsTerminated[_vault]) revert CannotAddTerminatedVault();
+        updateVault(_vault);
+        uint256 totalAllocPoint = (globalVaultsUpdates.length == 0) ? _allocPoint :
+        globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint - vaultInfo[_vault].allocPoint + _allocPoint;
+        if (globalVaultsUpdates.length > 0 &&
+            globalVaultsUpdates[globalVaultsUpdates.length-1].blockNumber == block.number) {
+            // already update in this block
+            globalVaultsUpdates[globalVaultsUpdates.length-1].totalAllocPoint = totalAllocPoint;
+        } else {
+            globalVaultsUpdates.push(VaultUpdate({
+                blockNumber: block.number,
+                totalAllocPoint: totalAllocPoint
+            }));
+        }
+
+        vaultInfo[_vault].allocPoint = _allocPoint;
+    }
+
+    function _commitUserBalance(
+        address _vault,
+        address _user,
+        uint256 _sharesChange,
+        bool _isDeposit
+    ) internal {
+        updateVault(_vault);
+
+        uint256 userShares = IERC20Upgradeable(_vault).balanceOf(_user);
+        uint256 rewardPerShare = vaultInfo[_vault].rewardPerShare;
+        unclaimedReward[_vault][_user] += userShares * rewardPerShare / REWARD_PRECISION - rewardDebt[_vault][_user];
+
+        if (_sharesChange != 0) {
+            if (_isDeposit) {
+                userShares += _sharesChange;
+            } else {
+                userShares -= _sharesChange;
+            }
+        }
+        rewardDebt[_vault][_user] = userShares * rewardPerShare / REWARD_PRECISION;
+    }
+
+    /**
+    * @notice Called by the vault to update a user claimable reward after deposit or withdraw.
+    * This call should never revert
+    * @param _user The user address to updare rewards for
+    * @param _sharesChange The user of shared the user deposited or withdrew
+    * @param _isDeposit Whether user deposited or withdrew
+    */
+    function commitUserBalance(address _user, uint256 _sharesChange, bool _isDeposit) external {
+        _commitUserBalance(msg.sender, _user, _sharesChange, _isDeposit);
+    }
+
+    /**
+    * @notice Transfer to the specified user their pending share of rewards.
+    * @param _vault The vault address
+    * @param _user The user address to claim for
+    */
+    function claimReward(address _vault, address _user) external {
```

Comment:
so anyone can call this for anyone now.. i thought anyway this should be in the vault and not the reward controller so maybe better to have it there and call this function from there? like this we will get both values from msg.sender

Reply:
it not just bad but also good that anyone can call this for anyone. It makes intergrations easier (someone can claim your rewards for you as a sservice)

Reply:
i don't see the use case but ok :) and don't you think it should be on the vault? since withdrawAndClaim and redeemAndClaim are there?

Reply:
this is the right place because all the relevant information is stored in this contract. Also, in this way users can claim their rewards also after we cahnage the rewardcontroller on the vault.